### PR TITLE
Added Clipnorm to the MC gradient descent

### PR DESCRIPTION
**Pull request by ecole41**
I have added clip norm to the MC method. This has a default of None but can be changed in the run card.

### CODE REVIEW

**Comment by ecole41** [comment]
> @ecole41 Hi Ella, I think this should work without touching the mc_fit at all, but by simply modifying (if needed) the optimiser, because then the optimiser is automatically passed to the mc_fit with the specific optimizer_settings you have set up.
> 
> So I think you just need to add the clipnorm feature in the optax provider and then it should work. The clipnorm is not a setting, so you are right that it should be an external parameter. We might want to have it under some control in the runcard though, are you setting it as a global variable at the moment?

Hi @LucaMantani , I agree. I have just edited this so that only the `optax_optimizer.py` file is changed such that clip norm can be controlled in the run card. This seems to work for me, is there anywhere else I should define this clip norm parameter?

**Comment by ecole41** [comment]
> Hi @ecole41 thanks for adding this. Could you please add some comments or examples on the reason or pros of having this feature? I guess you might have just used it in the fits you've been running?

Hi @comane, I have used this gradient clipping when running a Monte Carlo fit using the NNPDF n3fit model. I found that many replicas were finishing training with bad losses due to exploding gradients. This is maybe because of the increase in trainable parameters. After implementing this clipping, the replicas now all converge on good loss values around 1.

**Comment by ecole41** [comment]
> analytic_settings

I have just added this but it would be great if you were able to check if I have done this correctly. Thanks

**Review by LucaMantani — CHANGES_REQUESTED** [review]
Hi @ecole41 ,
it looks good but I realised that there is a slight problem which could be cool to fix. we now give optimizer_settings to opt:
```
opt(**optimizer_settings)
```
which gives us extreme flexibility because *any* setting that is implemented in optax or that will be implemented is allowed without having to code anything. 
However, if we check the known_keys this way, then we are losing this ability. If you look at the ns_settings, you will see that we solved this by doing nested dictionaries. I would propose to do the same and adapt then the optax_optimizer.py.

What i think would be nice to have is
```
optimizer_settings:
  optimizer: "adam"
  clipnorm: 1.0
  optimizer_hyperparams:
      learning_rate: 5e-4
```
This way everything is compact there, and the optimizer_hyperparams would be free to have different values. 

Also, could we have that clipnorm by default is not applied by setting it to None?

**Review by LucaMantani — APPROVED** [review]
I added the optimizer_settings to the docs and then testing it I realised that the signature of the function needed to be changed.

I directly took care of it. If you are happy with the changes I made, you can merge.
